### DISCUSSION TITLE
(Done) February 19th 2025 Meetup: Causal Profiling

### DISCUSSION THREAD

**Discussion by frenchy64** [madclj/madclj.com#11] [Announcements]
👍 to RSVP

https://madclj.com/

Wednesday, February 19th, 2025
6:30 PM to 9:00 PM CT

[StartingBlock Madison](https://www.google.com/maps/search/?api=1&query=43.081245%2C%20-89.374054)
821 E Washington Ave 2nd floor · Madison, WI

- https://github.com/clojure-goes-fast/clj-async-profiler
- https://github.com/Byte-Lab/JCoz
- https://github.com/tonsky/fast-edn/
- https://github.com/hugoduncan/criterium/

Ideas:
- causal profiler based on clj-async-profiler
- pure clojure causal profiler based on https://github.com/taoensso/tufte
- adapt JCoz to clojure https://github.com/Byte-Lab/JCoz/issues/23
- bench https://github.com/tonsky/fast-edn/ using JCoz
  - can we predict where clojure.edn reader is slow?
- summarize theory https://www.sigops.org/s/conferences/sosp/2015/current/2015-Monterey/printable/090-curtsinger.pdf
- existing discussion
  - https://www.reddit.com/r/Clojure/comments/kstuvl/anyone_tried_using_jcoz_java_causal_profiler_with/
  - https://clojureverse.org/t/avoiding-recursive-function-replacement-with-redefs-fn/5399/11

Organizer note:
- out of sparkling water

**Comment by frenchy64** [comment]
@madclj/meetup next meetup is about causal profiling. If you're keen to get involved or just want to hang out more, I've organized a planning lunch for [January 21st](https://github.com/orgs/madclj/discussions/22).

**Reply by brunchboy** [reply]
I was really impressed with JCoz so this seems intriguing!

**Reply by frenchy64** [reply]
Yes! I have more questions than answers but also intrigued.

**Comment by sivakusayan** [comment]
The research this originates from seems cool :) I'm a big fan of [Emery Berger's](https://emeryberger.com/) work in general, although I don't have the statistical background to appreciate his papers yet.

Only vaguely related, but this causal profiling stuff reminds me of a simulator called [uiCA](https://uops.info/uiCA.html), which tries to predict how well small blocks of x86-64 code will perform on Intel CPUs. Another similar tool is [llvm-mca](https://llvm.org/docs/CommandGuide/llvm-mca.html).

**Comment by frenchy64** [comment]
@madclj/meetup there will be no meetup on Feb 12th due to weather. We will instead meet on Feb 19th.

**Reply by brookeswanson** [reply]
Cool beans thanks for letting us know!

**Reply by brookeswanson** [reply]
Stay safe and warm everyone!

**Reply by brunchboy** [reply]
Ah, rats, I just got back from buying ingredients to bake olive cheese nuggets and amaretto cookies for tonight! Hopefully I’ll be able to do that next week instead. 😸

**Reply by frenchy64** [reply]
Oh no!!

**Reply by brunchboy** [reply]
It’s all good! I baked half-scale recipes today, and brought them to the condo social, and will make them fresh again next week. The amaretto cookies were a hit, people took extras home and left none for us, so I am looking forward to it. 😄

**Reply by brunchboy** [reply]
![IMG_2535](https://github.com/user-attachments/assets/c16eb787-4d95-4134-87d2-e3fdc439f785)

**Comment by ericnormand** [comment]
I'll be there with my 2 kids.

**Reply by brunchboy** [reply]
Nice, I just started cooking. 😄

**Reply by frenchy64** [reply]
Anyone planning to be late please drop a note here with an ETA. It's too cold to be left outside.

I'm planning to be at the door at 6:30pm.

**Reply by frenchy64** [reply]
Pizza is ordered (a bit less than usual). Grabbing more drinks and snacks, see you all there!

**Reply by brunchboy** [reply]
I am aiming to be there right at 6:30 but if the amaretto cookies need a little extra baking time I may be a couple minutes late, in which case I’ll text you.

**Reply by sivakusayan** [reply]
We'll be there at 6:50ish!

**Comment by frenchy64** [comment]
![signal-2025-02-19-215902_002](https://github.com/user-attachments/assets/9b276227-18e3-43d2-a30f-cc6d02b3a830)

- [Paper](https://arxiv.org/pdf/1608.03676v1)
- [coz](https://github.com/plasma-umass/coz)
- [Cannot reproduce plots](https://github.com/plasma-umass/coz/issues/241)
- [coz.h](https://github.com/plasma-umass/coz/blob/master/include/coz.h)
- [JCoz profiling for non-Java languages](https://github.com/Byte-Lab/JCoz/issues/23)
- [coz with jank (and futures)](https://github.com/frenchy64/jank/pull/8/files)
- [coz toy.cpp](https://github.com/plasma-umass/coz/blob/master/benchmarks/toy/toy.cpp#L21)
- [coz visualizer (`profile.coz` from jank below)](https://plasma-umass.org/coz/)

Save this as `profile.coz` and upload to visualizer:

```
startup	time=1739937085108150079
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=1.00	duration=441594961	selected-samples=59
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=77
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=0.00	duration=500971950	selected-samples=52
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=62
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.00	duration=500825363	selected-samples=120
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=81
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.05	duration=497261492	selected-samples=101
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=85
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.75	duration=414669445	selected-samples=115
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=81
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.50	duration=440094045	selected-samples=125
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=81
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.70	duration=431267364	selected-samples=105
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=87
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=0.00	duration=505283392	selected-samples=41
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=87
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.55	duration=447731070	selected-samples=113
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=86
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.45	duration=470618528	selected-samples=72
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=67
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.35	duration=472826768	selected-samples=83
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=73
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=0.60	duration=470850917	selected-samples=51
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=54
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.20	duration=487592277	selected-samples=63
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=49
runtime	time=6731119219
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	count=1273
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	count=691
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:29	count=3
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:30	count=2
samples	location=/usr/include/c++/14/bits/std_thread.h:173	count=5
startup	time=1739937403102455350
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.80	duration=405621387	selected-samples=119
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=80
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=0.00	duration=505300824	selected-samples=59
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=70
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.60	duration=446437742	selected-samples=93
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=77
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.95	duration=398179134	selected-samples=110
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=80
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.30	duration=478504138	selected-samples=74
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=64
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.20	duration=480671440	selected-samples=118
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=85
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.80	duration=428391381	selected-samples=95
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=70
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.30	duration=470754454	selected-samples=102
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=83
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.30	duration=471738024	selected-samples=105
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=86
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.60	duration=447774446	selected-samples=90
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=83
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.70	duration=441492564	selected-samples=84
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=80
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	speedup=0.80	duration=443835915	selected-samples=75
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=77
experiment	selected=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	speedup=0.90	duration=454214879	selected-samples=51
throughput-point	name=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:31	delta=38
runtime	time=6717131799
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:17	count=1271
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:21	count=669
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:29	count=1
samples	location=/home/ambrose/Projects/coz/benchmarks/toy/toy.cpp:30	count=1
samples	location=/usr/include/c++/14/bits/std_thread.h:173	count=2

```

<img width="907" alt="Screenshot 2025-02-19 at 10 07 28 PM" src="https://github.com/user-attachments/assets/554b9d78-84ed-48b6-85d4-281dce577f04" />
